cmd/disk-cache-fs.go

```diff
@@ -0,0 +1,424 @@
+		}
+		// This close will allow for locks to be synchronized on `fs.json`.
+		defer wlk.Close()
+		defer func() {
+			// Remove meta file when PutObject encounters any error
+			if retErr != nil {
+				tmpDir := pathJoin(fs.fsPath, minioMetaTmpBucket, fs.fsUUID)
+				fsRemoveMeta(bucketMetaDir, fsMetaPath, tmpDir)
+			}
+		}()
+	}
+
+	// Uploaded object will first be written to the temporary location which will eventually
+	// be renamed to the actual location. It is first written to the temporary location
+	// so that cleaning it up will be easy if the server goes down.
+	tempObj := mustGetUUID()
+
+	// Allocate a buffer to Read() from request body
+	bufSize := int64(readSizeV1)
+	if size := data.Size(); size > 0 && bufSize > size {
+		bufSize = size
+	}
+
+	buf := make([]byte, int(bufSize))
+	fsTmpObjPath := pathJoin(fs.fsPath, minioMetaTmpBucket, fs.fsUUID, tempObj)
+	bytesWritten, err := fsCreateFile(fsTmpObjPath, data, buf, data.Size())
+	if err != nil {
+		fsRemoveFile(fsTmpObjPath)
+		errorIf(err, "Failed to create object %s/%s", bucket, object)
+		return ObjectInfo{}, toObjectErr(err, bucket, object)
+	}
+	if fsMeta.Meta["etag"] == "" {
+		fsMeta.Meta["etag"] = hex.EncodeToString(data.MD5Current())
+	}
+	// Should return IncompleteBody{} error when reader has fewer
+	// bytes than specified in request header.
+	if bytesWritten < data.Size() {
+		fsRemoveFile(fsTmpObjPath)
+		return ObjectInfo{}, errors2.Trace(IncompleteBody{})
+	}
+
+	// Delete the temporary object in the case of a
+	// failure. If PutObject succeeds, then there would be
+	// nothing to delete.
+	defer fsRemoveFile(fsTmpObjPath)
+
+	// Entire object was written to the temp location, now it's safe to rename it to the actual location.
+	fsNSObjPath := pathJoin(fs.fsPath, bucket, object)
+	if err = fsRenameFile(fsTmpObjPath, fsNSObjPath); err != nil {
+		return ObjectInfo{}, toObjectErr(err, bucket, object)
+	}
+
+	if bucket != minioMetaBucket {
+		// Write FS metadata after a successful namespace operation.
+		if _, err = fsMeta.WriteTo(wlk); err != nil {
+			return ObjectInfo{}, toObjectErr(err, bucket, object)
+		}
+	}
+
+	// Stat the file to fetch timestamp, size.
+	fi, err := fsStatFile(pathJoin(fs.fsPath, bucket, object))
+	if err != nil {
+		return ObjectInfo{}, toObjectErr(err, bucket, object)
+	}
+	// Success.
+	return fsMeta.ToObjectInfo(bucket, object, fi), nil
+}
+
+// NewMultipartUpload - initialize a new multipart upload, returns a
+// unique id. The unique id returned here is of UUID form, for each
+// subsequent request each UUID is unique.
+//
+// Implements S3 compatible initiate multipart API. Operation here is identical
+// to fs backend implementation - with the exception that cache FS uses the uploadID
+// generated on the backend
+func (cfs *cacheFSObjects) NewMultipartUpload(bucket, object string, meta map[string]string, uploadID string) (string, error) {
+	if cfs.diskUsageHigh() {
+		select {
+		case cfs.purgeChan <- struct{}{}:
```

Comment:
Would this block a new upload from continuing until some disk space is cleared? Is it expected?

Reply:
Nvm - my bad above!